### PR TITLE
Adds the "Nanopocolypse" Event [DNM]

### PR DESCRIPTION
**Pull request by sawu-tg**
- Adds a Nanoswarm NPC that consumes everything it comes into contact with
- Nanoswarms have the following abilities:
- 1) They replace floor tiles they meet with their own nano-flooring
- 2) They consume power from SMES, APCs and cables
- 3) They consume stacks of objects (ie metal) to upgrade themselves.
- 4) They make machines and electronics malfunction via emagging
- 5) All unprotected humans they meet can be infected with an evil nanovirus, transforming them into psychotic cyborgs (now computer controlled).
- 6) They consume food and multiply, spreading the power evenly between themselves.

Nanoswarms can be combatted by the following procedures:
- A full-body protective suit, like a hardsuit, that exposes no flesh will prevent them from infecting you.
- EMP devices are especially effective against them, so load up on them.
- Disabling power in the area they are in will prevent them from becoming stronger
- Remove and store all food carefully so they do not multiply.
- Stem their flow by isolating them in containers, or surrounding them with Reinforced walls.
- Seek out the hives as soon as possible, and use EMP devices to destroy them (they are impervious to most other forms of harm).

Nanoswarms are controlled by overlords, picked from players who opt-in with the nanoswarm antag choice.

Overlords control swarms from a hive each, and have 5 abilities they can use in exchange for drained power.
- Spawn Nanoswarm: Creates a new nanoswarm child from the hive.
- Bolster: Spreads some of your power to your children, and upgrades them by 1 level.
- Psychic Screech: Emits a stunning screech that disables humans in your hive's vicinity.
- Call to Arms: Signals all of your children to hurry back to the hive.
- Call to Point: Signals all of your children to proceed to wherever your cursor located.

In addition to Overlord controlled Giga-hives, a number of regular, smaller hives will spawn. They automatically produce swarms at a slow rate, and serve as a distraction.

Profile for tile spreading (tested @ 25 swarms for about 10m):
cpu
nano1
nano2
nano3
space station 13 2015-04-17 214956

### CODE REVIEW

**Comment by Jordie0608** [comment]
I feel like this would be better as an event, probably admin-spawn only. Also changelog.

**Comment by WJohn** [comment]
I can feel the lag from here.

**Comment by Dorsisdwarf** [comment]
Finally, a use for flamethrowers

**Comment by Dicks1** [comment]
Will the psychotic cyborgs be npcs or player controlled?

**Comment by sawu-tg** [comment]
@Dicks1 Player.

File: code/modules/mob/living/simple_animal/hostile/nanoswarm.dm [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+			powerStored -= shared
+			NS.powerStored += shared
+
+		if(H)
+			var/ableToInject = 0
+			if(H.head && H.head.flags & THICKMATERIAL)
+				ableToInject++
+			if(H.wear_suit && H.wear_suit.flags & THICKMATERIAL)
+				ableToInject++
+			if(ableToInject < 2)
```
Comment (Dorsisdwarf) [inline code/modules/mob/living/simple_animal/hostile/nanoswarm.dm]: This makes no sense. Either AbleToInject should be lowered by having THICKMATERIAL suits, or it should be called "InjectResistance". This is not future-friendly.

File: code/modules/mob/living/simple_animal/hostile/nanoswarm.dm [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+/obj/structure/nanohive/proc/spawnNano()
+	if(health <= 0)
+		return
+	else
+		for(var/turf/T in oview(1))
+			if(curBots < maxBots)
+				curBots++
+				new /mob/living/simple_animal/hostile/nanoswarm(T)
+		spawn(interval)
+			spawnNano()
```
Comment (Dorsisdwarf) [inline code/modules/mob/living/simple_animal/hostile/nanoswarm.dm]: As soon as a single one of these is spawned, it's game over. This totally replenishes all nanoswarms every 3 seconds - You can't even kil them all that fast!

**Comment by Dorsisdwarf** [comment]
Also let's get some pictures in here. What does a serious infestation look like?

File: code/modules/mob/living/simple_animal/hostile/nanoswarm.dm (hunk 2 above)
Comment (sawu-tg) [inline code/modules/mob/living/simple_animal/hostile/nanoswarm.dm]: They take extra damage from EMP devices.

Liberal use of grenades and various EMPs is how they were designed to be taken down, not melee'd one by one.

**Comment by GunHog** [comment]
:+1: How does this affect borgs / AI / drones / station bots? Will they eventually die without power?

**Comment by sawu-tg** [comment]
@GunHog 

Good and bad, Nanoswarms only consume power, not destroy the power creators, so some power will be left for borgs and AI's to recharge.

Borgs are ignored when it comes to the transformation, and are mostly unaffected by the disabling beams that they fire (as well as drones).

If a nanoswarm has consumed too many base resources (like metal stacks), it will have a fairly heavy hit, so it can be a relative threat to an unprepared borg, and will probably floor drones.

The AI's turrets will target at, and kill, swarms very fast.

**Comment by PKPenguin321** [comment]
What's stopping an average player who wants antag status from seeking these out to become an evil cyborg?

**Comment by sawu-tg** [comment]
@PKPenguin321 The same thing that stops a player seeking out cult or rev for conversion, or seeking a wizard to construct them. Absolutely nothing, it would be a thing handle by administrators.

**Comment by Cheridan** [comment]
So... basically blob mode, then?

I don't think EMPs are common enough to use them as a counter for an enemy type in an entire gamemode. There's basically no way to fight it until miners bring back a load of uranium, then someone takes the uranium and grinds it down and makes a grenade from it.

Which basically sounds like a pain in the ass, so someone will inevitably want to add more EMP-producing items which will cause imbalances elsewhere.

**Comment by sawu-tg** [comment]
@Cheridan The picture may not be clear, it's kind of blob but kind of not. The central nests (the red pylons) spawn mobs who travel around being general shits (converting tiles, stealing power, destroying things).

EMPs are not the only way to kill them, but they are the most effective way, You could shoot it to death with lasers if it came to it. If people would like, I'll also make them vulnerable to fire.

**Comment by Anonus** [comment]
They should definitely be vulnerable to fire and burn damage, because otherwise there's only one EMP on the station at roundstart, and the other EMP needs mining.

**Comment by sawu-tg** [comment]
@Anonus I'll add that in as soon as possible then.

**Comment by Kromgar** [comment]
Yeah this mode should be a late game game ending event like aliens. You can't fight it early game you need lots of uranium and iron for emp nades

**Comment by GunHog** [comment]
Science can make Ion Carbines, you know. If you get mining, do not waste the uranium making grenades, make the guns instead. I am pretty sure Sec would cough up their firing pins with an enemy like that spreading.

**Comment by Dorsisdwarf** [comment]
Scientists really will be hard-pressed to research, what with being dragged into nanswarms by the borged greytide as the infection reaches supersonic speed

**Comment by francinum** [comment]
Wasn't there a slime core that created an EMP?

**Comment by sawu-tg** [comment]
@Kromgar This is negated with the addition of fire weakness, anyone with access to flamethrower materials can fight them.

@Dorsisdwarf See above, that being said, targeting ranks is natural in some game modes (cult, monkey etc)

@francinum  Several slime cores are useful against them. Orange slime and plasma creates fire, Oil slime and plasma explodes, Yellow slime and blood EMPS.

**Comment by Dicks1** [comment]
Why not have the mobs ai controlled like blob zombies to stop players from rushing in to die?

**Comment by sawu-tg** [comment]
@Dicks1 I wanted to give players a certain level of interaction as an antag in this game mode. Purposely sacrificing yourself just to get antag is against the rules afaik and should be dealt with by admins.

**Comment by Steelpoint** [comment]
I think the mechanic of humans being turned into evil cyborgs needs to be tossed out.

Giving free antag status to people will result in people flinging themselves at the Nanomachines, no matter how good the admins are doing their job. It should be akin to the blob where the person dies and loses all control.

Also on a art note, it would be cool to see the Nanozombies look less like Cyborgs and more like a meld of flesh and machine.

Anyway, thislooks pretty cool.

**Comment by PKPenguin321** [comment]
Forum feedback thread: https://tgstation13.org/phpBB/viewtopic.php?f=10&t=3416

**Comment by Jalleo** [comment]
@sawu-tg have you thought of instead of player controlled it interacts like a blob spore to a human thats dead and takes control?
This requires a few play tests before giving it a final say really its a tough one theory is not what it will be.

**Comment by Incomptinence** [comment]
Just thought I would note that nanoborgs use the syndicate module here. 

What that essentially means is grenade launchers everywhere any temporary rwall barrier just being blown to girders and eaten. Like if all blob zombies had gyrojets or something.

**Comment by sawu-tg** [comment]
I am working on a total rework of most of the mechanics, please do not merge until it's done.

**Comment by sawu-tg** [comment]
A number of things changed, but this is now merge ready.
Players no longer converted into controllable robots, instead a tiny chance to spawn evil robot NPCs from them on hit.
"Overlords" are chosen at the start of the game to control a hive each.

**Comment by Jordie0608** [comment]
After running a test server for a bit, sawu has said he's going to rework this a bit, DNM until then.

**Comment by optimumtact** [comment]
Make this a round, it's like a more interesting blob

**Comment by Jordie0608** [comment]
It was a gamemode when we tested it.

**Comment by sawu-tg** [comment]
I don't have the time to work on this right now, I'll remake it and PR it some other time.

**Comment by bruhlookatthisdood** [comment]
are you gonna remake it yet?

**Comment by sawu-tg** [comment]
> are you gonna remake it yet?

It's been almost over 7 years. I've lost the original source code and have moved onto my own projects from tg-station.

How the heck did you even find a 7 year old PR to comment on?

**Comment by bruhlookatthisdood** [comment]
> How the heck did you even find a 7 year old PR to comment on?

i... i dont know.

**Comment by optimumtact** [comment]
kvTSRlWAReekYHdyQsebjJxR

**Pull request closed without merging**